src/hotspot/share/prims/jvmtiEnvBase.cpp

```diff
@@ -1485,14 +1485,28 @@ JvmtiEnvBase::get_object_monitor_usage(JavaThread* calling_thread, jobject objec
     // this object has a heavyweight monitor
     nWant = mon->contentions(); // # of threads contending for monitor
     nWait = mon->waiters();     // # of threads in Object.wait()
-    ret.waiter_count = nWant + nWait;
-    ret.notify_waiter_count = nWait;
   } else {
     // this object has a lightweight monitor
-    ret.waiter_count = 0;
-    ret.notify_waiter_count = 0;
   }
 
+  ResourceMark rm(current_thread);
+  GrowableArray<JavaThread*>* wantList = Threads::get_pending_threads(tlh.list(), nWant + nWait, (address)mon);
+  nWant = wantList->length();
+
+  if (mon != nullptr) {
+    ObjectWaiter *waiter = mon->first_waiter();
+    for (int i = 0; i < nWait; i++) {
+      if (waiter == nullptr || (i != 0 && waiter == mon->first_waiter())) {
+        // robustness: the waiting list has gotten smaller
```

Comment:
We are at a safepoint so I don't see how the wait list can shrink. I initially thought perhaps a waiter could timeout, but the code that does the timed park is wrapped in ` ThreadBlockInVMPreprocess` which will block at a safepoint if one is active.

Reply:
Thank you for the question.
The `nWait` count we got from the `mon->waiters()` can include the threads that are re-entering the monitor after being notified. Here we are correcting the actual number of the waiting threads by excluding those re-entering the monitor. The comment need an update to explain it.